### DISCUSSION TITLE
Is there a way to rapidly calculate the mAP for each image?

### DISCUSSION THREAD

**Discussion by istranic** [Lightning-AI/torchmetrics#1073] [CompVision] [answered]
Hi,

I'd like to log the mAP for each image in my object detection dataset, and as far as I can tell, the only way to do that is to run `metric(preds, target)` after each image. This is very slow, so I'm wondering if there's a way for `metric.compute()` to return metrics per image without having to call it individually for each image?

**Comment by Borda** [comment, marked as answer]
The latest TorchMetrics documentation shows that mAP is calculated using `MeanAveragePrecision`, with predictions and targets provided as lists of dictionaries—each dictionary corresponding to one image. By default, calling `compute()` returns only the global dataset metrics, such as overall mAP, mAR, and optionally per-class results (with `class_metrics=True`). There is currently no setting that returns separate mAP values for each image directly—mAP is inherently a dataset-level metric and not naturally defined per image. The returned metrics are aggregated over all images.
If you need to obtain mAP for each image, you do need to call the metric (either via the functional interface or by instantiating the metric and updating with one image at a time) individually for each image, as you suspected. This can be slow with large datasets.
